From 5fe13334ac93c4d4cd06938b8b6e128266ce4b1c Mon Sep 17 00:00:00 2001
From: Michal Sekletar <msekleta@redhat.com>
Date: Tue, 27 Jan 2015 14:55:14 +0100
Subject: [PATCH] Fix compilation failure on RHEL6 where TP_STATUS_VLAN_VALID
 is not available

---
 pcap-linux.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/pcap-linux.c b/pcap-linux.c
index e028f09..61d3c27 100644
--- a/pcap-linux.c
+++ b/pcap-linux.c
@@ -1642,9 +1642,11 @@ pcap_read_packet(pcap_t *handle, pcap_handler callback, u_char *userdata)
 			tag->vlan_tci = htons(aux->tp_vlan_tci);
 
                         /* store vlan tci to bpf_aux_data struct for userland bpf filter */
-#if defined(TP_STATUS_VLAN_VALID)
                         aux_data.vlan_tag = htons(aux->tp_vlan_tci) & 0x0fff;
+#if defined(TP_STATUS_VLAN_VALID)
                         aux_data.vlan_tag_present = (aux->tp_status & TP_STATUS_VLAN_VALID);
+#else
+                        aux_data.vlan_tag_present = (aux->tp_vlan_tci != 0);
 #endif
 			packet_len += VLAN_TAG_LEN;
 		}
@@ -4080,7 +4082,7 @@ pcap_read_linux_mmap(pcap_t *handle, int max_packets, pcap_handler callback,
                         struct bpf_aux_data aux_data;
 
                         aux_data.vlan_tag = h.h2->tp_vlan_tci;
-                        aux_data.vlan_tag_present = (h.h2->tp_vlan_tci || (h.h2->tp_status & TP_STATUS_VLAN_VALID));
+                        aux_data.vlan_tag_present = (h.h2->tp_vlan_tci != 0);
 
                         if (bpf_filter_with_aux_data(handle->fcode.bf_insns, bp, tp_len, tp_snaplen, &aux_data) == 0)
                                 goto skip;
-- 
1.8.3.1

