From 83402c1f11b9fa88b953a8167b1ceaca9c394eda Mon Sep 17 00:00:00 2001
From: Michal Sekletar <msekleta@redhat.com>
Date: Wed, 28 Jan 2015 10:09:55 +0100
Subject: [PATCH] Fix issues discovered by coverity

---
 pcap-linux.c | 4 +++-
 pcap/bpf.h   | 2 +-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/pcap-linux.c b/pcap-linux.c
index 61d3c27..c874002 100644
--- a/pcap-linux.c
+++ b/pcap-linux.c
@@ -144,6 +144,7 @@ static const char rcsid[] _U_ =
 #include "pcap-int.h"
 #include "pcap/sll.h"
 #include "pcap/vlan.h"
+#include "pcap/bpf.h"
 
 /*
  * If PF_PACKET is defined, we can use {SOCK_RAW,SOCK_DGRAM}/PF_PACKET
@@ -2971,7 +2972,8 @@ activate_new(pcap_t *handle)
 	int			err = 0;
 	struct packet_mreq	mr;
 #ifdef SO_BPF_EXTENSIONS
-	int			bpf_extensions, len;
+	int			bpf_extensions;
+	socklen_t		len;
 #endif
 
 	/*
diff --git a/pcap/bpf.h b/pcap/bpf.h
index 63e18a2..d1ba17f 100644
--- a/pcap/bpf.h
+++ b/pcap/bpf.h
@@ -1334,7 +1334,7 @@ struct bpf_aux_data {
 #if __STDC__ || defined(__cplusplus)
 extern int bpf_validate(const struct bpf_insn *, int);
 extern u_int bpf_filter(const struct bpf_insn *, const u_char *, u_int, u_int);
-extern u_int bpf_filter1(const struct bpf_insn *, const u_char *, u_int, u_int, const struct bpf_aux_data *);
+extern u_int bpf_filter_with_aux_data(const struct bpf_insn *, const u_char *, u_int, u_int, const struct bpf_aux_data *);
 #else
 extern int bpf_validate();
 extern u_int bpf_filter();
-- 
1.8.3.1

