From dab69e6e147dcdcae5c8e4d777652bdead8d592a Mon Sep 17 00:00:00 2001
From: Michal Sekletar <msekleta@redhat.com>
Date: Wed, 28 Jan 2015 13:34:21 +0100
Subject: [PATCH] Initialize len before calling getsockopt

---
 pcap-linux.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/pcap-linux.c b/pcap-linux.c
index c874002..1375ffa 100644
--- a/pcap-linux.c
+++ b/pcap-linux.c
@@ -3309,6 +3309,7 @@ activate_new(pcap_t *handle)
 	 * (XXX - what if we need the special code but it's not supported
 	 * by the OS?  Is that possible?)
 	 */
+	len = sizeof(bpf_extensions);
 	if (getsockopt(sock_fd, SOL_SOCKET, SO_BPF_EXTENSIONS, &bpf_extensions, &len) == 0) {
 		if (bpf_extensions >= SKF_AD_VLAN_TAG_PRESENT) {
 			/*
-- 
1.8.3.1

