diff --git a/Makefile.in b/Makefile.in
index 22fd0af..23db41c 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -166,7 +166,9 @@ MAN3PCAP_EXPAND = \
 	pcap_list_tstamp_types.3pcap.in \
 	pcap_open_dead.3pcap.in \
 	pcap_open_offline.3pcap.in \
-	pcap_set_tstamp_type.3pcap.in
+	pcap_set_tstamp_type.3pcap.in \
+	pcap_set_tstamp_precision.3pcap.in \
+	pcap_get_tstamp_precision.3pcap.in
 
 MAN3PCAP_NOEXPAND = \
 	pcap_activate.3pcap \
