From cc8994ccdab5b885a6f42276d84fd6eaf6c1491c Mon Sep 17 00:00:00 2001
From: Michal Sekletar <msekleta@redhat.com>
Date: Tue, 27 Aug 2013 19:23:07 +0200
Subject: [PATCH 4/5] linux: remove dead code

---
 pcap-linux.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/pcap-linux.c b/pcap-linux.c
index 8b82ca7..ee1ef05 100644
--- a/pcap-linux.c
+++ b/pcap-linux.c
@@ -427,8 +427,6 @@ pcap_create_interface(const char *device, char *ebuf)
 	handle->tstamp_precision_count = 2;
 	handle->tstamp_type_list = malloc(2 * sizeof(u_int));
 	if (handle->tstamp_type_list == NULL) {
-		if (handle->tstamp_type_list != NULL)
-			free(handle->tstamp_type_list);
 		free(handle);
 		return NULL;
 	}
-- 
1.8.3.1

