diff -up libpcap-1.5.0-PRE-GIT/pcap-linux.c.s390 libpcap-1.5.0-PRE-GIT/pcap-linux.c
--- libpcap-1.5.0-PRE-GIT/pcap-linux.c.s390	2013-08-01 17:34:32.000485575 +0200
+++ libpcap-1.5.0-PRE-GIT/pcap-linux.c	2013-08-01 17:42:39.947139763 +0200
@@ -3143,6 +3143,10 @@ activate_new(pcap_t *handle)
 				handle->linktype = DLT_LINUX_SLL;
 		}
 
+		/* Hack to make things work on s390 ctc interfaces */
+		if (strncmp("ctc", device, 3) == 0)
+			handle->linktype = DLT_EN10MB;
+
 		handlep->ifindex = iface_get_id(sock_fd, device,
 		    handle->errbuf);
 		if (handlep->ifindex == -1) {
